irohad/torii/processor/impl/transaction_processor_impl.cpp

```diff
@@ -15,106 +15,109 @@
  * limitations under the License.
  */
 
-#include <crypto/hash.hpp>
-#include <endpoint.pb.h>
+#include "torii/processor/transaction_processor_impl.hpp"
+
 #include <iostream>
-#include <model/transaction_response.hpp>
-#include <torii/processor/transaction_processor_impl.hpp>
 #include <utility>
 
+#include <boost/range/adaptor/filtered.hpp>
+#include <boost/range/algorithm/for_each.hpp>
+
+#include "crypto/hash.hpp"
+#include "endpoint.pb.h"
+#include "model/transaction_response.hpp"
+
 namespace iroha {
   namespace torii {
 
-    using validation::StatelessValidator;
     using model::TransactionResponse;
+    using Status = model::TransactionResponse::Status;
     using network::PeerCommunicationService;
+    using validation::StatelessValidator;
 
     TransactionProcessorImpl::TransactionProcessorImpl(
         std::shared_ptr<PeerCommunicationService> pcs,
-        std::shared_ptr<StatelessValidator> validator)
-        : pcs_(std::move(pcs)), validator_(std::move(validator)) {
+        std::shared_ptr<StatelessValidator> validator,
+        std::shared_ptr<MstProcessor> mst_processor)
+        : pcs_(std::move(pcs)),
+          validator_(std::move(validator)),
+          mst_processor_(std::move(mst_processor)) {
       log_ = logger::log("TxProcessor");
 
       // insert all txs from proposal to proposal set
       pcs_->on_proposal().subscribe([this](model::Proposal proposal) {
         for (const auto &tx : proposal.transactions) {
           proposal_set_.insert(hash(tx).to_string());
-          TransactionResponse response;
-          response.tx_hash = hash(tx).to_string();
-          response.current_status =
-              TransactionResponse::STATELESS_VALIDATION_SUCCESS;
-          notifier_.get_subscriber().on_next(
-              std::make_shared<model::TransactionResponse>(response));
+          notify(hash(tx).to_string(), Status::STATELESS_VALIDATION_SUCCESS);
         }
       });
 
       // move commited txs from proposal to candidate map
-      pcs_->on_commit().subscribe([this](
-          rxcpp::observable<model::Block> blocks) {
+      pcs_->on_commit().subscribe([this](auto &&blocks) {
         blocks.subscribe(
             // on next..
-            [this](model::Block block) {
-              for (const auto &tx : block.transactions) {
-                if (this->proposal_set_.count(hash(tx).to_string())) {
-                  proposal_set_.erase(hash(tx).to_string());
-                  candidate_set_.insert(hash(tx).to_string());
-                  TransactionResponse response;
-                  response.tx_hash = hash(tx).to_string();
-                  response.current_status =
-                      model::TransactionResponse::STATEFUL_VALIDATION_SUCCESS;
-                  notifier_.get_subscriber().on_next(
-                      std::make_shared<model::TransactionResponse>(response));
-                }
-              }
+            [this](auto &&block) {
```

Comment:
I don't understand what these changes do it integration MST PR. I guess more confident to make it in separate PR.

Reply:
Sorry, for that inconvenience
Don't really want to copy-paste similar code that's why this refactoring is made (though boost is not that necessary)